### ISSUE TITLE
Should be able to directly sign with sighashes

### ISSUE THREAD

**Issue by apoelstra** [rust-bitcoin/rust-bitcoin#2823]
Related to https://github.com/rust-bitcoin/rust-secp256k1/issues/700

If you have a `SegwitV0Hash`, or any other kind of sighash, there should be an easy/obvious way to produce a signature on this. Maybe the hash itself should have a `.sign` method that takes a secp context(?) and a public key.

Also, we should check our signmessage API -- while this is basically deprecated upstream, we should give it a tiny bit of love by giving it its own sighash constructor which prepends the `Bitcoin Signed Message:` prefix and having the signing API look uniform with the signing API for the other sighash types.

**Comment by tcharding** [comment]
I did this ages ago in https://github.com/rust-bitcoin/rust-bitcoin/pull/2008, what's the priority - I can revive it.

**Comment by apoelstra** [comment]
Not very high. Just, if you find yourself sweeping for issues to address, it's here.

**Comment by tcharding** [comment]
Also, @lamfab originally tried to do this in https://github.com/rust-bitcoin/rust-bitcoin/pull/1932 and I took over, now I realize this I feel a bit bad that we didn't resolve it before last release (assuming he had a usecase for it).

**Comment by sanket1729** [comment]
One other potential footgun I observed in our signing/verify APIs is that they take the hash value as the input. For a ECDSA signature, you want to make sure hashing is done as a part of sign ECDSA API, but we have users explicitly create a sighash that is later passed onto sign API for signing.

@apoelstra do you know if the libsecp ECDSA API does the hashing internally or can I _recover_ a message hash that satisfies satoshi's public key?

**Comment by apoelstra** [comment]
@sanket1729 no, the API does not do hashing internally. So you can indeed recover a message hash (and a few years back somebody, I think actually Craig Wright, actually did this and posted it on Reddit).

**Comment by Kixunil** [comment]
> I don't mind having a backdoor where you can go to/from raw bytes but maybe we should consider dropping that for the sighash types?

I think this is related to tagging hashes with `KnownPreimage` - those should be only constructable with hashed data or by explicitly acknowledging the danger.

**Comment by apoelstra** [comment]
Yep, agreed. I think this is exactly the same issue.
